docs/rfc/rc-005-event-system.rst

```diff
@@ -0,0 +1,121 @@
+to receive notifications after a series of given events, such as new blocks,
+new transactions, validator set changes, as well as intermediate events during
+consensus. Because the event system is very cross cutting, the behavior and
+performance of the event publication and subscription system has huge impacts
+for all of Tendermint.
+
+The subscription service is exposed over the RPC interface, but also powers
+the indexing (e.g. to an external database,) and is the mechanism by which
+`BroadcastTxCommit` is able to wait for transactions to land in a block.
+
+The current pubsub mechanism relies on a couple of buffered channels,
+primarily between all event creators and subscribers, but also for each
+subscription. The result of this design is that, in some situations with the
+right collection of slow subscription consumers the event system can put
+backpressure on the consensus state machine and message gossiping in the
+network, thereby causing nodes to lag.
+
+Improvements
+~~~~~~~~~~~~
+
+The current system relies on implicit queues built by the buffered channels,
+and though threadsafe, can force all activity within Tendermint to serialize,
+which does not need to happen. Additionally, timeouts for subscription
+consumers related to the implementation of the RPC layer, may complicate the
+use of the system.
+
+References
+~~~~~~~~~~
+
+- Legacy Implementation
+  - `publication of events <https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L333-L345>`_ 
+  - `send operation <https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L489-L527>`_ 
+  - `send loop <https://github.com/tendermint/tendermint/blob/master/libs/pubsub/pubsub.go#L381-L402>`_
+- Related RFCs 
+  - `RFC 002: IPC Ecosystem <./rfc-002-ipc-ecosystem.md>`_ 
+  - `RFC 003: Performance Questions <./rfc-003-performance-questions.md>`_ 
+
+Discussion
+----------
+
+Changes to Published Events
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+As part of this process, the Tendermint team should do a study of the existing
+event types and ensure that there are viable production use cases for
+subscriptions to all event types. Instinctively it seems plausible that some
+of the events may not be useable outside of tendermint, (e.g. ``TimeoutWait``
+or ``NewRoundStep``) and it might make sense to remove them. Certainly, it
+would be good to make sure that we don't maintain infrastructure for unused or
+un-useful message indefinitely.
+
+Blocking Subscription
+~~~~~~~~~~~~~~~~~~~~~
+
+The blocking subscription mechanism makes it possible to have *send*
+operations into the subscription channel be un-buffered (the event processing
+channel is still buffered.) In the blocking case, events from one subscription
+can block processing that event for other non-blocking subscriptions. The main
+case, it seems for blocking subscriptions is ensuring that a transaction has
+been committed to a block for ``BroadcastTxCommit``. Removing blocking
```

Comment:
If BroadcastTxCommit is really the only use case for "blocking" subscriptions, then we can definitely find a better solution (even a webhook would probably be better than freezing consensus). But if there are other use cases, this is another case where we should define the service levels more explicitly.

I'm also not convinced BroadcastTxCommit is a service we should keep baked into the consensus node. If we have a way for a client to subscribe and catch up on an event stream reliably, then the waiting logic can be moved out of the core into the client (or a library the client calls). Broadcast a transaction in the usual way, then go do other things while you wait for the appropriate event to fire. (We might need to support better filtering. Possibly a proxy would make sense, assuming only certain nodes need to handle subscriber traffic)

Reply:
Yes, that might be sufficient for the common case. It might be worth doing this as a UX affordance.

That said: I think we have to deal with the more general issue of how a client discovers when/if its Tx lands in the chain, even if a connection fails or it takes longer than 10 seconds. My intuition is that BroadcastTxCommit should probably not exist, or should be built on top of a more reliable mechanism (e.g., inject the transaction and {monitor events, watch a light client, other}) and exposed as library functionality. I do _not_ think it belongs in the primary consensus API.

To really solve that we'll need input from users about how they're using this now, and how it does(n't) serve their needs. It's possible it might not implicate the event system at all, though I suspect it probably will.